physx/src/controller.rs

```diff
@@ -0,0 +1,103 @@
+#![warn(clippy::all)]
+#![warn(rust_2018_idioms)]
+
+use crate::px_type::*;
+use crate::rigid_dynamic::RigidDynamic;
+
+use glam::Vec3;
+use physx_macros::*;
+
+use physx_sys::{
+    phys_PxCreateControllerManager, PxCapsuleControllerDesc, PxCapsuleControllerDesc_delete,
+    PxCapsuleControllerDesc_isValid, PxCapsuleControllerDesc_new_alloc, PxController,
+    PxControllerDesc, PxControllerManager, PxControllerManager_createController_mut,
+    PxController_getActor, PxController_getPosition, PxController_release_mut,
+    PxController_setPosition_mut, PxExtendedVec3, PxMaterial, PxScene,
+};
+
+#[physx_type]
+impl ControllerDesc {}
+
+#[physx_type]
+impl ControllerManager {
+    pub fn new(scene: *mut PxScene) -> Self {
+        let ptr = unsafe { phys_PxCreateControllerManager(scene, false) };
+        ControllerManager { ptr }
+    }
+
+    pub fn create_controller(&self, desc: &mut ControllerDesc) -> Controller {
+        let controller =
+            unsafe { PxControllerManager_createController_mut(self.ptr, desc.get_raw_mut()) };
+        Controller::new(controller)
+    }
+}
+
+#[physx_type(inherit = "ControllerDesc")]
+impl CapsuleControllerDesc {
+    pub fn new(
+        height: f32,
+        radius: f32,
+        step_offset: f32,
+        material: *mut PxMaterial,
+    ) -> Result<Self, String> {
+        unsafe {
+            let c = PxCapsuleControllerDesc_new_alloc();
+            (*c).height = height;
+            (*c).radius = radius;
+            (*c).stepOffset = step_offset;
+            (*c).material = material;
+
+            if PxCapsuleControllerDesc_isValid(c) {
+                Ok(CapsuleControllerDesc { ptr: c })
+            } else {
+                Err(format!(
```

Comment:
Returning strings as errors is suboptimal, since it's hard to match on them on the consumer side. The most common approach is to return an enum that implements std::error::Error, like this:

```rust
pub mod controller {
    pub enum Error {
        InvalidDesc { height: f32, radius: f32, offset: f32 }
    }
    impl std::error::Error for Error {
        ...
    }

    type Result<T> = std::result::Result<T, Error>;
}

fn maybe_succeeds() -> controller::Result<Something> { ... }
```

There's a crate called `thiserror` that's commonly used to make this a bit nicer: https://docs.rs/thiserror/1.0.11/thiserror/. It should be quite straightforward to convert your String errors use this library instead, but I'll happily provide more info or help out if you want.

Also, we should free the pointer in the failure branch as well, so we don't leak it.

Reply:
Ok, added the error enum